lib/routemaster/api_client_circuit.rb

```diff
@@ -0,0 +1,43 @@
+require 'circuitbox'
+require 'routemaster/errors'
+module Routemaster
+  class APIClientCircuit
+    def initialize(url)
+      url = URI.parse(url) unless url.is_a? URI
+      @circuit_name = url.host.downcase
+    end
+
+    def call(&block)
+      if enabled?
+        begin
+          return circuit.run!(&block)
+        rescue Circuitbox::ServiceFailureError => e
+          raise e.original
+        end
+      else
+        return block.call
+      end
+    end
+
+    private
+
+    def enabled?
+      ENV.fetch('ROUTEMASTER_ENABLE_API_CLIENT_CIRCUIT', 'NO') =~ /\A(YES|TRUE|ON|1)\Z/i
+    end
+
+    def circuit
+      Circuitbox.circuit(@circuit_name, {
+        sleep_window: configuration_setting(@circuit_name, 'ROUTEMASTER_CIRCUIT_BREAKER_SLEEP_WINDOW', 60).to_i,
+        time_window: configuration_setting(@circuit_name, 'ROUTEMASTER_CIRCUIT_BREAKER_TIME_WINDOW', 120).to_i,
+        volume_threshold: configuration_setting(@circuit_name, 'ROUTEMASTER_CIRCUIT_BREAKER_VOLUME_THRESHOLD', 50).to_i,
+        error_threshold:  configuration_setting(@circuit_name, 'ROUTEMASTER_CIRCUIT_BREAKER_ERROR_THRESHOLD', 50).to_i,
+        cache: Moneta.new(:Redis, backend: Config.cache_redis),
+        exceptions: [Routemaster::Errors::FatalResource, Faraday::TimeoutError]
```

Comment:
I just noticed this, but `FatalResource` includes the response codes 407..500.

A lot of those 4xx codes should _not_ indicate server issues (should not contribute to opening the circuit) and there are a number of unchecked 5xx response codes that _should_ indicate server issues (503, 504).

Reply:
I've changed the list so 500 + and 408 return FatalResource

Reply:
I think it'd be wise to address this in another PR. This has been running for a week without any accidental open circuits